### PR TITLE
Free applicative functors

### PR DESCRIPTION
**Pull request by runarorama**

### CODE REVIEW

**Comment by runarorama** [comment]
Review by @ekmett would be good

File: core/src/main/scala/scalaz/Frap.scala [hunk 1]
```diff
@@ -0,0 +1,118 @@
+package scalaz
+
+/**
+ * Free applicative functors. Less expressive than free monads, but more
+ * flexible to inspect and interpret.
+ */
+sealed trait Frap[F[_],A] {
```
Comment (jedws) [inline core/src/main/scala/scalaz/Frap.scala]: `FreeAp`?

File: core/src/main/scala/scalaz/Frap.scala [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+   * Not tail-recursive unless `G` is a free monad.
+   */
+  def foldMap[G[_]:Applicative](f: F ~> G): G[A] =
+    this match {
+      case Pure(x) => Applicative[G].pure(x)
+      case x@Ap() => Applicative[G].ap(f(x.v()))(x.k() foldMap f)
+    }
+
+  /** Provides access to the first instruction of this program, if present */
+  def para[B](pure: A => B, ap: ({type λ[α] = (F[α], Frap[F, α => A])})#λ ~>
```
Comment (jedws) [inline core/src/main/scala/scalaz/Frap.scala]: is scalaz going to use kind-projector? would really help…

**Comment by folone** [comment]
Nice! I tried doing [something similar](https://github.com/folone/free.scala/blob/master/src/main/scala/free/applicative/Applicative.scala) some time ago, but failed miserably :)

File: core/src/main/scala/scalaz/Frap.scala (hunk 2 above)
Comment (runarorama) [inline core/src/main/scala/scalaz/Frap.scala]: That would be awesome. I think you just volunteered to go through the codebase and replace type lambdas with that.

**Comment by jdegoes** [comment]
:open_mouth: :+1:

File: core/src/main/scala/scalaz/Frap.scala (hunk 2 above)
Comment (jedws) [inline core/src/main/scala/scalaz/Frap.scala]: bastard! https://github.com/scalaz/scalaz/pull/875

Comment (runarorama) [inline core/src/main/scala/scalaz/Frap.scala]: Wait, what? Oh, can KP not support f[x] = (x,x) ?

Comment (jedws) [inline core/src/main/scala/scalaz/Frap.scala]: I can't find this on GH, replying by email. Yes, KP can do that via the
type function syntax, eg: `λ[α => (α, α)]`

**Pull request merged**